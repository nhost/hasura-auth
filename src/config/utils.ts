// * Helpers for casting environment variables

export const castStringEnv = (envVar: string, defaultValue = ''): string => {
  return process.env[envVar] ? (process.env[envVar] as string) : defaultValue;
};

export const castBooleanEnv = (
  envVar: string,
  defaultValue = false
): boolean => {
  return process.env[envVar]
    ? process.env[envVar]?.toLowerCase() === 'true'
    : defaultValue;
};

export const castIntEnv = (envVar: string, defaultValue: number): number => {
  return process.env[envVar]
    ? parseInt(process.env[envVar] as string, 10)
    : defaultValue;
};

export const castStringArrayEnv = (
  envVar: string,
  defaultValue: string[] = []
): string[] => {
  return process.env[envVar]?.length
    ? (process.env[envVar] as string).split(',').map((field) => field.trim())
    : defaultValue;
};

export const castObjectEnv = <T extends Record<string, unknown>>(
  envVar: string,
  defaultValue: T = {} as T
): T => {
  const env = process.env[envVar];
  if (env) {
    return JSON.parse(env);
  } else return defaultValue;
};
