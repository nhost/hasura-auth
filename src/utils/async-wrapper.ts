import { NextFunction, Response, Request } from 'express';

/**
 * This wrapper function sends any route errors to `next()`.
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function asyncWrapper(fn: any) {
  return function (req: Request, res: Response, next: NextFunction): void {
    fn(req, res, next).catch(next);
  };
}
