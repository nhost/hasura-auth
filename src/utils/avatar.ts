import { url } from 'gravatar';
import { ENV } from './env';

export const getGravatarUrl = (email?: string) => {
  if (ENV.AUTH_GRAVATAR_ENABLED && email) {
    return url(email, {
      r: ENV.AUTH_GRAVATAR_RATING,
      protocol: 'https',
      default: ENV.AUTH_GRAVATAR_DEFAULT,
    });
  }
};
