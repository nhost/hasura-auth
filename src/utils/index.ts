export * from './redirect';
export * from './password';
export * from './user';
export * from './qr-code';
export * from './avatar';
export * from './otp';
export * from './gql-sdk';
export * from './ticket';
export * from './tokens';
export * from './jwt';
export * from './env';
export * from './hasura';
export * from './async-wrapper';
