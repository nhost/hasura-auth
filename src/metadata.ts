import axios from 'axios';
import { logger } from './logger';
import { ENV } from './utils/env';

/**
 * https://hasura.io/docs/latest/graphql/core/api-reference/schema-metadata-api/relationship.html
 * Here we are using the schema-metadata-api to track the relationships between auth tables
 **/

interface Table {
  name: string;
  schema: string;
}

interface TableArgs {
  source?: string;
  table: Table;
}

type TableConfig = {
  custom_name?: string;
  identifier?: string;
  custom_root_fields?: {
    select?: string;
    select_by_pk?: string;
    select_aggregate?: string;
    insert?: string;
    insert_one?: string;
    update?: string;
    update_by_pk?: string;
    delete?: string;
    delete_by_pk?: string;
  };
  custom_column_names?: {
    [key: string]: string;
  };
};

type TrackTableArgs = TableArgs & {
  configuration?: TableConfig;
};
type UntrackTableArgs = TableArgs & {
  cascade?: boolean;
};

type TableCustomisationArgs = TableArgs & {
  configuration?: TableConfig;
};

type DropRelationshipArgs = TableArgs & {
  relationship: string;
};

type CreateRelationshipArgs = TableArgs & {
  name: string;
  using: {
    foreign_key_constraint_on:
      | {
          table: Table;
          columns: string[];
        }
      | string[];
  };
};

export const runMetadataRequest = async (args: { type: string; args: {} }) => {
  await axios.post(
    ENV.HASURA_GRAPHQL_GRAPHQL_URL.replace('/v1/graphql', '/v1/metadata'),
    args,
    {
      headers: {
        'x-hasura-admin-secret': ENV.HASURA_GRAPHQL_ADMIN_SECRET,
      },
    }
  );
};

// https://hasura.io/docs/latest/graphql/core/api-reference/schema-metadata-api/table-view.html#track-table-v2
export const trackTable = async (args: TrackTableArgs) => {
  try {
    await runMetadataRequest({
      type: 'pg_track_table',
      args,
    });
  } catch (error: any) {
    if (error.response.data.code !== 'already-tracked') {
      logger.error(error);
      throw new Error(`Error tracking table ${args.table.name}`);
    } else {
      logger.debug(`Table ${args.table.name} already tracked`);
    }
  }
};

// https://hasura.io/docs/latest/graphql/core/api-reference/schema-metadata-api/table-view.html#untrack-table
export const untrackTable = async (args: UntrackTableArgs) => {
  try {
    await runMetadataRequest({
      type: 'pg_untrack_table',
      args,
    });
  } catch (error: any) {
    logger.error(error);
    throw new Error(`Error untracking table ${args.table.name}`);
  }
};

export const reloadMetadata = async (
  args: {
    reload_remote_schemas?: boolean;
    reload_sources?: boolean;
    recreate_event_triggers?: boolean;
  } = {}
) => {
  await runMetadataRequest({
    type: 'reload_metadata',
    args,
  });
};
export const setTableCustomization = async (args: TableCustomisationArgs) => {
  logger.info(`Set table customization for ${args.table.name}`);

  try {
    await runMetadataRequest({
      type: 'pg_set_table_customization',
      args,
    });
  } catch (error: any) {
    logger.error(error);
    throw new Error('error setting customization for table ' + args.table.name);
  }
};

export const createObjectRelationship = async (
  args: CreateRelationshipArgs
) => {
  logger.info(`Set object relationship ${args.name} for ${args.table.name}`);
  try {
    await runMetadataRequest({
      type: 'pg_create_object_relationship',
      args,
    });
  } catch (error: any) {
    if (error.response.data.code !== 'already-exists') {
      throw new Error(
        `Error creating object relationship for table ${args.table.name}`
      );
    } else {
      logger.debug(
        `Object relationship ${args.name} on table ${args.table.name} is already created`
      );
    }
  }
};

export const createArrayRelationship = async (args: CreateRelationshipArgs) => {
  logger.info(`Create array relationship ${args.name} for ${args.table.name}`);
  try {
    await runMetadataRequest({
      type: 'pg_create_array_relationship',
      args,
    });
  } catch (error: any) {
    if (error.response.data.code !== 'already-exists') {
      throw new Error(
        `Error creating array relationship for table ${args.table.name}`
      );
    }
    logger.debug(
      `Array relationship ${args.name} on table ${args.table.name} is already created`
    );
  }
};

export const dropRelationship = async (args: DropRelationshipArgs) => {
  logger.info(`Drop relationship ${args.relationship} for ${args.table.name}`);
  try {
    await runMetadataRequest({
      type: 'pg_drop_relationship',
      args,
    });
  } catch (error: any) {
    if (error.response.data.code !== 'already-exists') {
      throw new Error(
        `Error dropping relationship for table ${args.table.name}`
      );
    } else {
      logger.debug(
        `Object relationship ${args.relationship} on table ${args.table.name} is already created`
      );
    }
  }
};

export const applyMetadata = async (): Promise<void> => {
  logger.info('Applying metadata...');

  logger.debug('Reloading metadata...');
  await reloadMetadata();
  logger.debug('Metadata reloaded');

  try {
    // track
    await trackTable({
      source: 'default',
      table: {
        schema: 'auth',
        name: 'provider_requests',
      },
    });
    await trackTable({
      source: 'default',
      table: {
        schema: 'auth',
        name: 'refresh_tokens',
      },
    });
    await trackTable({
      source: 'default',
      table: {
        schema: 'auth',
        name: 'roles',
      },
    });
    await trackTable({
      source: 'default',
      table: {
        schema: 'auth',
        name: 'user_providers',
      },
    });
    await trackTable({
      source: 'default',
      table: {
        schema: 'auth',
        name: 'user_roles',
      },
    });
    await trackTable({
      source: 'default',
      table: {
        schema: 'auth',
        name: 'users',
      },
    });
    await trackTable({
      source: 'default',
      table: {
        schema: 'auth',
        name: 'providers',
      },
    });
    await trackTable({
      source: 'default',
      table: {
        schema: 'auth',
        name: 'user_authenticators',
      },
    });

    // customization
    await setTableCustomization({
      source: 'default',
      table: {
        schema: 'auth',
        name: 'users',
      },
      configuration: {
        custom_name: 'users',
        custom_root_fields: {
          select: 'users',
          select_by_pk: 'user',
          select_aggregate: 'usersAggregate',
          insert: 'insertUsers',
          insert_one: 'insertUser',
          update: 'updateUsers',
          update_by_pk: 'updateUser',
          delete: 'deleteUsers',
          delete_by_pk: 'deleteUser',
        },
        custom_column_names: {
          // * See: https://github.com/hasura/graphql-engine/issues/8020
          // id: 'id',
          created_at: 'createdAt',
          updated_at: 'updatedAt',
          last_seen: 'lastSeen',
          // * See: https://github.com/hasura/graphql-engine/issues/8020
          // disabled: 'disabled',
          display_name: 'displayName',
          avatar_url: 'avatarUrl',
          // * See: https://github.com/hasura/graphql-engine/issues/8020
          // locale: 'locale',
          // email: 'email',
          phone_number: 'phoneNumber',
          password_hash: 'passwordHash',
          email_verified: 'emailVerified',
          phone_number_verified: 'phoneNumberVerified',
          new_email: 'newEmail',
          otp_method_last_used: 'otpMethodLastUsed',
          otp_hash: 'otpHash',
          otp_hash_expires_at: 'otpHashExpiresAt',
          default_role: 'defaultRole',
          is_anonymous: 'isAnonymous',
          totp_secret: 'totpSecret',
          active_mfa_type: 'activeMfaType',
          // * See: https://github.com/hasura/graphql-engine/issues/8020
          // ticket: 'ticket',
          ticket_expires_at: 'ticketExpiresAt',
          webauthn_current_challenge: 'currentChallenge',
        },
      },
    });

    await setTableCustomization({
      source: 'default',
      table: {
        schema: 'auth',
        name: 'user_roles',
      },
      configuration: {
        custom_name: 'authUserRoles',
        custom_root_fields: {
          select: 'authUserRoles',
          select_by_pk: 'authUserRole',
          select_aggregate: 'authUserRolesAggregate',
          insert: 'insertAuthUserRoles',
          insert_one: 'insertAuthUserRole',
          update: 'updateAuthUserRoles',
          update_by_pk: 'updateAuthUserRole',
          delete: 'deleteAuthUserRoles',
          delete_by_pk: 'deleteAuthUserRole',
        },
        custom_column_names: {
          // * See: https://github.com/hasura/graphql-engine/issues/8020
          // id: 'id',
          created_at: 'createdAt',
          user_id: 'userId',
          // * See: https://github.com/hasura/graphql-engine/issues/8020
          // role: 'role',
        },
      },
    });
    await setTableCustomization({
      source: 'default',
      table: {
        schema: 'auth',
        name: 'user_providers',
      },
      configuration: {
        custom_name: 'authUserProviders',
        custom_root_fields: {
          select: 'authUserProviders',
          select_by_pk: 'authUserProvider',
          select_aggregate: 'authUserProvidersAggregate',
          insert: 'insertAuthUserProviders',
          insert_one: 'insertAuthUserProvider',
          update: 'updateAuthUserProviders',
          update_by_pk: 'updateAuthUserProvider',
          delete: 'deleteAuthUserProviders',
          delete_by_pk: 'deleteAuthUserProvider',
        },
        custom_column_names: {
          // * See: https://github.com/hasura/graphql-engine/issues/8020
          // id: 'id',
          created_at: 'createdAt',
          updated_at: 'updatedAt',
          user_id: 'userId',
          access_token: 'accessToken',
          refresh_token: 'refreshToken',
          provider_id: 'providerId',
          provider_user_id: 'providerUserId',
        },
      },
    });
    await setTableCustomization({
      source: 'default',
      table: {
        schema: 'auth',
        name: 'providers',
      },
      configuration: {
        custom_name: 'authProviders',
        custom_root_fields: {
          select: 'authProviders',
          select_by_pk: 'authProvider',
          select_aggregate: 'authProvidersAggregate',
          insert: 'insertAuthProviders',
          insert_one: 'insertAuthProvider',
          update: 'updateAuthProviders',
          update_by_pk: 'updateAuthProvider',
          delete: 'deleteAuthProviders',
          delete_by_pk: 'deleteAuthProvider',
        },
        custom_column_names: {
          // * See: https://github.com/hasura/graphql-engine/issues/8020
          // id: 'id',
        },
      },
    });
    await setTableCustomization({
      source: 'default',
      table: {
        schema: 'auth',
        name: 'refresh_tokens',
      },
      configuration: {
        custom_name: 'authRefreshTokens',
        custom_root_fields: {
          select: 'authRefreshTokens',
          select_by_pk: 'authRefreshToken',
          select_aggregate: 'authRefreshTokensAggregate',
          insert: 'insertAuthRefreshTokens',
          insert_one: 'insertAuthRefreshToken',
          update: 'updateAuthRefreshTokens',
          update_by_pk: 'updateAuthRefreshToken',
          delete: 'deleteAuthRefreshTokens',
          delete_by_pk: 'deleteAuthRefreshToken',
        },
        custom_column_names: {
          refresh_token: 'refreshToken',
          created_at: 'createdAt',
          expires_at: 'expiresAt',
          user_id: 'userId',
        },
      },
    });
    await setTableCustomization({
      source: 'default',
      table: {
        schema: 'auth',
        name: 'roles',
      },
      configuration: {
        custom_name: 'authRoles',
        custom_root_fields: {
          select: 'authRoles',
          select_by_pk: 'authRole',
          select_aggregate: 'authRolesAggregate',
          insert: 'insertAuthRoles',
          insert_one: 'insertAuthRole',
          update: 'updateAuthRoles',
          update_by_pk: 'updateAuthRole',
          delete: 'deleteAuthRoles',
          delete_by_pk: 'deleteAuthRole',
        },
        custom_column_names: {
          // * See: https://github.com/hasura/graphql-engine/issues/8020
          // role: 'role',
        },
      },
    });
    await setTableCustomization({
      source: 'default',
      table: {
        schema: 'auth',
        name: 'provider_requests',
      },
      configuration: {
        custom_name: 'authProviderRequests',
        custom_root_fields: {
          select: 'authProviderRequests',
          select_by_pk: 'authProviderRequest',
          select_aggregate: 'authProviderRequestsAggregate',
          insert: 'insertAuthProviderRequests',
          insert_one: 'insertAuthProviderRequest',
          update: 'updateAuthProviderRequests',
          update_by_pk: 'updateAuthProviderRequest',
          delete: 'deleteAuthProviderRequests',
          delete_by_pk: 'deleteAuthProviderRequest',
        },
        custom_column_names: {
          // * See: https://github.com/hasura/graphql-engine/issues/8020
          // id: 'id',
          // redirect_url: 'redirectUrl',
        },
      },
    });
    await setTableCustomization({
      source: 'default',
      table: {
        schema: 'auth',
        name: 'user_authenticators',
      },
      configuration: {
        custom_name: 'authUserAuthenticators',
        custom_root_fields: {
          select: 'authUserAuthenticators',
          select_by_pk: 'authUserAuthenticator',
          select_aggregate: 'authUserAuthenticatorsAggregate',
          insert: 'insertAuthUserAuthenticators',
          insert_one: 'insertAuthUserAuthenticator',
          update: 'updateAuthUserAuthenticators',
          update_by_pk: 'updateAuthUserAuthenticator',
          delete: 'deleteAuthUserAuthenticators',
          delete_by_pk: 'deleteAuthUserAuthenticator',
        },
        custom_column_names: {
          // * See: https://github.com/hasura/graphql-engine/issues/8020
          // id: 'id',
          user_id: 'userId',
          credential_id: 'credentialId',
          credential_public_key: 'credentialPublicKey',
        },
      },
    });

    await createObjectRelationship({
      source: 'default',
      table: {
        schema: 'auth',
        name: 'user_providers',
      },
      name: 'user',
      using: {
        foreign_key_constraint_on: ['user_id'],
      },
    });
    await createArrayRelationship({
      source: 'default',
      table: {
        schema: 'auth',
        name: 'users',
      },
      name: 'userProviders',
      using: {
        foreign_key_constraint_on: {
          table: {
            schema: 'auth',
            name: 'user_providers',
          },
          columns: ['user_id'],
        },
      },
    });

    await createObjectRelationship({
      source: 'default',
      table: {
        schema: 'auth',
        name: 'user_providers',
      },
      name: 'provider',
      using: {
        foreign_key_constraint_on: ['provider_id'],
      },
    });
    await createArrayRelationship({
      source: 'default',
      table: {
        schema: 'auth',
        name: 'providers',
      },
      name: 'userProviders',
      using: {
        foreign_key_constraint_on: {
          table: {
            schema: 'auth',
            name: 'user_providers',
          },
          columns: ['provider_id'],
        },
      },
    });

    await createObjectRelationship({
      source: 'default',
      table: {
        schema: 'auth',
        name: 'user_roles',
      },
      name: 'user',
      using: {
        foreign_key_constraint_on: ['user_id'],
      },
    });

    await createArrayRelationship({
      source: 'default',
      table: {
        schema: 'auth',
        name: 'roles',
      },
      name: 'userRoles',
      using: {
        foreign_key_constraint_on: {
          table: {
            schema: 'auth',
            name: 'user_roles',
          },
          columns: ['role'],
        },
      },
    });

    await createObjectRelationship({
      source: 'default',
      table: {
        schema: 'auth',
        name: 'users',
      },
      name: 'defaultRoleByRole',
      using: {
        foreign_key_constraint_on: ['default_role'],
      },
    });
    await createArrayRelationship({
      source: 'default',
      table: {
        schema: 'auth',
        name: 'roles',
      },
      name: 'usersByDefaultRole',
      using: {
        foreign_key_constraint_on: {
          table: {
            schema: 'auth',
            name: 'users',
          },
          columns: ['default_role'],
        },
      },
    });

    await createArrayRelationship({
      source: 'default',
      table: {
        schema: 'auth',
        name: 'users',
      },
      name: 'roles',
      using: {
        foreign_key_constraint_on: {
          table: {
            schema: 'auth',
            name: 'user_roles',
          },
          columns: ['user_id'],
        },
      },
    });

    await createObjectRelationship({
      source: 'default',
      table: {
        schema: 'auth',
        name: 'user_roles',
      },
      name: 'user',
      using: {
        foreign_key_constraint_on: ['user_id'],
      },
    });

    await createObjectRelationship({
      source: 'default',
      table: {
        schema: 'auth',
        name: 'user_roles',
      },
      name: 'roleByRole',
      using: {
        foreign_key_constraint_on: ['role'],
      },
    });

    await createObjectRelationship({
      source: 'default',
      table: {
        schema: 'auth',
        name: 'refresh_tokens',
      },
      name: 'user',
      using: {
        foreign_key_constraint_on: ['user_id'],
      },
    });
    await createArrayRelationship({
      source: 'default',
      table: {
        schema: 'auth',
        name: 'users',
      },
      name: 'refreshTokens',
      using: {
        foreign_key_constraint_on: {
          table: {
            schema: 'auth',
            name: 'refresh_tokens',
          },
          columns: ['user_id'],
        },
      },
    });

    await createObjectRelationship({
      source: 'default',
      table: {
        schema: 'auth',
        name: 'user_authenticators',
      },
      name: 'user',
      using: {
        foreign_key_constraint_on: ['user_id'],
      },
    });
    await createArrayRelationship({
      source: 'default',
      table: {
        schema: 'auth',
        name: 'users',
      },
      name: 'authenticators',
      using: {
        foreign_key_constraint_on: {
          table: {
            schema: 'auth',
            name: 'user_authenticators',
          },
          columns: ['user_id'],
        },
      },
    });
  } finally {
    logger.debug('Reloading metadata...');
    await reloadMetadata();
    logger.debug('Metadata reloaded');
  }
  logger.info('Metadata applied');
};
