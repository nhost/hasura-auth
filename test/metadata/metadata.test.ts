import { hasuraAuthMetadataPatch } from '@/metadata';
import { HasuraMetadataV3, patchMetadataObject } from '@/utils';
import { EMPTY_METADATA } from './empty-metadata';
import { NHOST_PROJECT_METADATA } from './nhost-project-metadata';

describe('metadata operations', () => {
  it('should patch metadata from an empty project', () => {
    const metadata: HasuraMetadataV3 = { ...EMPTY_METADATA };
    patchMetadataObject(metadata, hasuraAuthMetadataPatch);
    expect(metadata).toMatchSnapshot();
  });

  it('should not change the metadata when hasura-auth already applied its metadata', () => {
    const metadata: HasuraMetadataV3 = { ...NHOST_PROJECT_METADATA };
    patchMetadataObject(metadata, hasuraAuthMetadataPatch);
    expect(metadata).toMatchObject(NHOST_PROJECT_METADATA);
  });

  describe('fix broken metadata', () => {
    it('should track back an untracked table', () => {
      const metadata: HasuraMetadataV3 = { ...NHOST_PROJECT_METADATA };
      metadata.sources[0].tables = metadata.sources[0].tables.filter(
        ({ table: { name, schema } }) =>
          !(name === 'users' && schema === 'auth')
      );
      patchMetadataObject(metadata, hasuraAuthMetadataPatch);
      expect(metadata).toMatchObject(NHOST_PROJECT_METADATA);
    });

    it('should track back a missing relationship', () => {
      const metadata: HasuraMetadataV3 = { ...NHOST_PROJECT_METADATA };
      metadata.sources[0].tables
        .find(
          ({ table: { name, schema } }) => name === 'users' && schema === 'auth'
        )
        ?.array_relationships?.pop();
      patchMetadataObject(metadata, hasuraAuthMetadataPatch);
      expect(metadata).toMatchObject(NHOST_PROJECT_METADATA);
    });
  });
});
