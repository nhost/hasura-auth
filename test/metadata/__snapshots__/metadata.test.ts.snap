// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`metadata operations should patch metadata from an empty project 1`] = `
Object {
  "sources": Array [
    Object {
      "configuration": Object {
        "connection_info": Object {
          "database_url": Object {
            "from_env": "HASURA_GRAPHQL_DATABASE_URL",
          },
          "isolation_level": "read-committed",
          "pool_settings": Object {
            "connection_lifetime": 600,
            "idle_timeout": 180,
            "max_connections": 50,
            "retries": 1,
          },
          "use_prepared_statements": true,
        },
      },
      "kind": "postgres",
      "name": "default",
      "tables": Array [
        Object {
          "configuration": Object {
            "custom_column_names": Object {
              "id": "id",
              "options": "options",
            },
            "custom_name": "authProviderRequests",
            "custom_root_fields": Object {
              "delete": "deleteAuthProviderRequests",
              "delete_by_pk": "deleteAuthProviderRequest",
              "insert": "insertAuthProviderRequests",
              "insert_one": "insertAuthProviderRequest",
              "select": "authProviderRequests",
              "select_aggregate": "authProviderRequestsAggregate",
              "select_by_pk": "authProviderRequest",
              "update": "updateAuthProviderRequests",
              "update_by_pk": "updateAuthProviderRequest",
            },
          },
          "table": Object {
            "name": "provider_requests",
            "schema": "auth",
          },
        },
        Object {
          "array_relationships": Array [
            Object {
              "name": "refreshTokens",
              "using": Object {
                "foreign_key_constraint_on": Object {
                  "column": "type",
                  "table": Object {
                    "name": "refresh_tokens",
                    "schema": "auth",
                  },
                },
              },
            },
          ],
          "configuration": Object {
            "custom_name": "authRefreshTokenTypes",
            "custom_root_fields": Object {
              "delete": "deleteAuthRefreshTokenTypes",
              "delete_by_pk": "deleteAuthRefreshTokenType",
              "insert": "insertAuthRefreshTokenTypes",
              "insert_one": "insertAuthRefreshTokenType",
              "select": "authRefreshTokenTypes",
              "select_aggregate": "authRefreshTokenTypesAggregate",
              "select_by_pk": "authRefreshTokenType",
              "update": "updateAuthRefreshTokenTypes",
              "update_by_pk": "updateAuthRefreshTokenType",
            },
          },
          "is_enum": true,
          "table": Object {
            "name": "refresh_token_types",
            "schema": "auth",
          },
        },
        Object {
          "configuration": Object {
            "column_config": Object {},
            "custom_column_names": Object {
              "created_at": "createdAt",
              "expires_at": "expiresAt",
              "refresh_token_hash": "refreshTokenHash",
              "user_id": "userId",
            },
            "custom_name": "authRefreshTokens",
            "custom_root_fields": Object {
              "delete": "deleteAuthRefreshTokens",
              "delete_by_pk": "deleteAuthRefreshToken",
              "insert": "insertAuthRefreshTokens",
              "insert_one": "insertAuthRefreshToken",
              "select": "authRefreshTokens",
              "select_aggregate": "authRefreshTokensAggregate",
              "select_by_pk": "authRefreshToken",
              "update": "updateAuthRefreshTokens",
              "update_by_pk": "updateAuthRefreshToken",
            },
          },
          "delete_permissions": Array [
            Object {
              "permission": Object {
                "filter": Object {
                  "_and": Array [
                    Object {
                      "user_id": Object {
                        "_eq": "X-Hasura-User-Id",
                      },
                    },
                    Object {
                      "type": Object {
                        "_eq": "pat",
                      },
                    },
                  ],
                },
              },
              "role": "user",
            },
          ],
          "object_relationships": Array [
            Object {
              "name": "user",
              "using": Object {
                "foreign_key_constraint_on": "user_id",
              },
            },
          ],
          "select_permissions": Array [
            Object {
              "permission": Object {
                "columns": Array [
                  "id",
                  "created_at",
                  "expires_at",
                  "metadata",
                  "type",
                  "user_id",
                ],
                "filter": Object {
                  "user_id": Object {
                    "_eq": "X-Hasura-User-Id",
                  },
                },
              },
              "role": "user",
            },
          ],
          "table": Object {
            "name": "refresh_tokens",
            "schema": "auth",
          },
        },
        Object {
          "array_relationships": Array [
            Object {
              "name": "userRoles",
              "using": Object {
                "foreign_key_constraint_on": Object {
                  "column": "role",
                  "table": Object {
                    "name": "user_roles",
                    "schema": "auth",
                  },
                },
              },
            },
            Object {
              "name": "usersByDefaultRole",
              "using": Object {
                "foreign_key_constraint_on": Object {
                  "column": "default_role",
                  "table": Object {
                    "name": "users",
                    "schema": "auth",
                  },
                },
              },
            },
          ],
          "configuration": Object {
            "custom_column_names": Object {
              "role": "role",
            },
            "custom_name": "authRoles",
            "custom_root_fields": Object {
              "delete": "deleteAuthRoles",
              "delete_by_pk": "deleteAuthRole",
              "insert": "insertAuthRoles",
              "insert_one": "insertAuthRole",
              "select": "authRoles",
              "select_aggregate": "authRolesAggregate",
              "select_by_pk": "authRole",
              "update": "updateAuthRoles",
              "update_by_pk": "updateAuthRole",
            },
          },
          "table": Object {
            "name": "roles",
            "schema": "auth",
          },
        },
        Object {
          "configuration": Object {
            "custom_column_names": Object {
              "access_token": "accessToken",
              "created_at": "createdAt",
              "id": "id",
              "provider_id": "providerId",
              "provider_user_id": "providerUserId",
              "refresh_token": "refreshToken",
              "updated_at": "updatedAt",
              "user_id": "userId",
            },
            "custom_name": "authUserProviders",
            "custom_root_fields": Object {
              "delete": "deleteAuthUserProviders",
              "delete_by_pk": "deleteAuthUserProvider",
              "insert": "insertAuthUserProviders",
              "insert_one": "insertAuthUserProvider",
              "select": "authUserProviders",
              "select_aggregate": "authUserProvidersAggregate",
              "select_by_pk": "authUserProvider",
              "update": "updateAuthUserProviders",
              "update_by_pk": "updateAuthUserProvider",
            },
          },
          "object_relationships": Array [
            Object {
              "name": "user",
              "using": Object {
                "foreign_key_constraint_on": "user_id",
              },
            },
            Object {
              "name": "provider",
              "using": Object {
                "foreign_key_constraint_on": "provider_id",
              },
            },
          ],
          "table": Object {
            "name": "user_providers",
            "schema": "auth",
          },
        },
        Object {
          "configuration": Object {
            "custom_column_names": Object {
              "created_at": "createdAt",
              "id": "id",
              "role": "role",
              "user_id": "userId",
            },
            "custom_name": "authUserRoles",
            "custom_root_fields": Object {
              "delete": "deleteAuthUserRoles",
              "delete_by_pk": "deleteAuthUserRole",
              "insert": "insertAuthUserRoles",
              "insert_one": "insertAuthUserRole",
              "select": "authUserRoles",
              "select_aggregate": "authUserRolesAggregate",
              "select_by_pk": "authUserRole",
              "update": "updateAuthUserRoles",
              "update_by_pk": "updateAuthUserRole",
            },
          },
          "object_relationships": Array [
            Object {
              "name": "user",
              "using": Object {
                "foreign_key_constraint_on": "user_id",
              },
            },
            Object {
              "name": "roleByRole",
              "using": Object {
                "foreign_key_constraint_on": "role",
              },
            },
          ],
          "table": Object {
            "name": "user_roles",
            "schema": "auth",
          },
        },
        Object {
          "array_relationships": Array [
            Object {
              "name": "userProviders",
              "using": Object {
                "foreign_key_constraint_on": Object {
                  "column": "user_id",
                  "table": Object {
                    "name": "user_providers",
                    "schema": "auth",
                  },
                },
              },
            },
            Object {
              "name": "roles",
              "using": Object {
                "foreign_key_constraint_on": Object {
                  "column": "user_id",
                  "table": Object {
                    "name": "user_roles",
                    "schema": "auth",
                  },
                },
              },
            },
            Object {
              "name": "refreshTokens",
              "using": Object {
                "foreign_key_constraint_on": Object {
                  "column": "user_id",
                  "table": Object {
                    "name": "refresh_tokens",
                    "schema": "auth",
                  },
                },
              },
            },
            Object {
              "name": "securityKeys",
              "using": Object {
                "foreign_key_constraint_on": Object {
                  "column": "user_id",
                  "table": Object {
                    "name": "user_security_keys",
                    "schema": "auth",
                  },
                },
              },
            },
          ],
          "configuration": Object {
            "custom_column_names": Object {
              "active_mfa_type": "activeMfaType",
              "avatar_url": "avatarUrl",
              "created_at": "createdAt",
              "default_role": "defaultRole",
              "disabled": "disabled",
              "display_name": "displayName",
              "email": "email",
              "email_verified": "emailVerified",
              "id": "id",
              "is_anonymous": "isAnonymous",
              "last_seen": "lastSeen",
              "locale": "locale",
              "new_email": "newEmail",
              "otp_hash": "otpHash",
              "otp_hash_expires_at": "otpHashExpiresAt",
              "otp_method_last_used": "otpMethodLastUsed",
              "password_hash": "passwordHash",
              "phone_number": "phoneNumber",
              "phone_number_verified": "phoneNumberVerified",
              "ticket": "ticket",
              "ticket_expires_at": "ticketExpiresAt",
              "totp_secret": "totpSecret",
              "updated_at": "updatedAt",
              "webauthn_current_challenge": "currentChallenge",
            },
            "custom_name": "users",
            "custom_root_fields": Object {
              "delete": "deleteUsers",
              "delete_by_pk": "deleteUser",
              "insert": "insertUsers",
              "insert_one": "insertUser",
              "select": "users",
              "select_aggregate": "usersAggregate",
              "select_by_pk": "user",
              "update": "updateUsers",
              "update_by_pk": "updateUser",
            },
          },
          "object_relationships": Array [
            Object {
              "name": "defaultRoleByRole",
              "using": Object {
                "foreign_key_constraint_on": "default_role",
              },
            },
          ],
          "table": Object {
            "name": "users",
            "schema": "auth",
          },
        },
        Object {
          "array_relationships": Array [
            Object {
              "name": "userProviders",
              "using": Object {
                "foreign_key_constraint_on": Object {
                  "column": "provider_id",
                  "table": Object {
                    "name": "user_providers",
                    "schema": "auth",
                  },
                },
              },
            },
          ],
          "configuration": Object {
            "custom_column_names": Object {
              "id": "id",
            },
            "custom_name": "authProviders",
            "custom_root_fields": Object {
              "delete": "deleteAuthProviders",
              "delete_by_pk": "deleteAuthProvider",
              "insert": "insertAuthProviders",
              "insert_one": "insertAuthProvider",
              "select": "authProviders",
              "select_aggregate": "authProvidersAggregate",
              "select_by_pk": "authProvider",
              "update": "updateAuthProviders",
              "update_by_pk": "updateAuthProvider",
            },
          },
          "table": Object {
            "name": "providers",
            "schema": "auth",
          },
        },
        Object {
          "configuration": Object {
            "custom_column_names": Object {
              "credential_id": "credentialId",
              "credential_public_key": "credentialPublicKey",
              "id": "id",
              "user_id": "userId",
            },
            "custom_name": "authUserSecurityKeys",
            "custom_root_fields": Object {
              "delete": "deleteAuthUserSecurityKeys",
              "delete_by_pk": "deleteAuthUserSecurityKey",
              "insert": "insertAuthUserSecurityKeys",
              "insert_one": "insertAuthUserSecurityKey",
              "select": "authUserSecurityKeys",
              "select_aggregate": "authUserSecurityKeysAggregate",
              "select_by_pk": "authUserSecurityKey",
              "update": "updateAuthUserSecurityKeys",
              "update_by_pk": "updateAuthUserSecurityKey",
            },
          },
          "object_relationships": Array [
            Object {
              "name": "user",
              "using": Object {
                "foreign_key_constraint_on": "user_id",
              },
            },
          ],
          "table": Object {
            "name": "user_security_keys",
            "schema": "auth",
          },
        },
      ],
    },
  ],
  "version": 3,
}
`;
